### PR TITLE
[Onnx] - refactoring serialization of ONNX initializers to be name-based

### PR DESCRIPTION
**Pull request by spandantiwari**
Currently, serialization of model parameters in ONNX export depends on the order in which they are stored in a container (`list` on Python side and `std::vector` on C++ side). This has worked fine till now, but if we need to do any pass on that graph that mutates the parameter list, then strictly order-based serialization may not work.

This PR is the first in a set to bring in more passes (such as constant folding) related to ONNX export. This PR lays the groundwork by moving the serialization in ONNX export from order-based to name based approach, which is more amenable to some of the passes. 

@houseroad - As discussed this change uses a map for export, and removes the code from `export.cpp` that relies on the order to compute initializer names.

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
Overall, the idea looks good to me. Could you fix the build issue?

**Comment by spandantiwari** [comment]
@houseroad - yes, working on it.

File: torch/csrc/jit/export.cpp [hunk 1]
```diff
@@ -303,16 +303,10 @@ void EncoderBase::EncodeBlock(
       EncodeBlock(false_g, node->blocks()[1]);
     }
   }
-  auto num_initializers = initializers.size();
-  AT_ASSERT(block->inputs().size() >= num_initializers);
```
[the remark this answers is not in the document]
Comment (spandantiwari) [reply torch/csrc/jit/export.cpp]: Agreed. Bringing it back.

**Comment by spandantiwari** [comment]
Submitted a fix for the build issue.

**Comment by spandantiwari** [comment]
Seeing test failure(s). Investigating...

**Comment by spandantiwari** [comment]
A couple of ONNX export tests in `test_jit.py` were failing because now the order of initializers could possibly differ. Did a sanity check that the model is indeed the same as before and checked that the model outputs is still the same. Fixed the test by updating the `.expect` files for them.

@houseroad - I also saw some unexpected SIGINT type failure in the last build (possibly similar to your `reserve_param` PR CI build). Waiting to see if that shows up in this build or not.

**Comment by spandantiwari** [comment]
@houseroad - I am seeing that SIGFPE type failure again. Could you please take a look and suggest what can be done to fix this. Thanks.

**Comment by spandantiwari** [comment]
Rebased to master.

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -303,16 +305,15 @@ void EncoderBase::EncodeBlock(
...
-  size_t inputs_count = block->inputs().size() - num_initializers;
-  for (auto& tensor : initializers) {
-    // TODO: stop using positions to determine which initializers
-    // match to which inputs
-    std::string name = graph_proto->input(inputs_count++).name();
+  AT_ASSERT(block->inputs().size() >= initializers.size());
+  // Copying initializers to into an (ordered) stdd::map so that
+  // initializers always gets written into ONNX graph in the same
```
Comment (houseroad) [inline torch/csrc/jit/export.cpp]: Typo: stdd

File: torch/csrc/jit/export.cpp [hunk 3]
```diff
@@ -303,16 +305,15 @@ void EncoderBase::EncodeBlock(
...
-  for (auto& tensor : initializers) {
-    // TODO: stop using positions to determine which initializers
-    // match to which inputs
-    std::string name = graph_proto->input(inputs_count++).name();
+  AT_ASSERT(block->inputs().size() >= initializers.size());
+  // Copying initializers to into an (ordered) stdd::map so that
+  // initializers always gets written into ONNX graph in the same 
+  // order. Mostly done for testing purpose.
+  for (auto& name_tensor_pair : 
+        std::map<std::string, at::Tensor>(initializers.begin(), initializers.end())) {
```
Comment (houseroad) [inline torch/csrc/jit/export.cpp]: Which tests will fail? We probably can change it. This hack seems not a good way to solve the problem

**Review by houseroad — COMMENTED** [review]
If we would like to reserve the order, we may try using orderedmap (in python)

File: torch/csrc/jit/export.cpp (hunk 2 above)
Comment (spandantiwari) [reply torch/csrc/jit/export.cpp]: Done.

File: torch/csrc/jit/export.cpp (hunk 3 above)
Comment (spandantiwari) [reply torch/csrc/jit/export.cpp]: Updated based on discussion to std::map.

**Comment by spandantiwari** [comment]
Updated the export method to use `std::map` to get deterministic order of serialization of the initializers during ONNX export for testing purposes.

**Review by houseroad — APPROVED** [review]
Looks good! This is a great change.

**Comment by spandantiwari** [comment]
@houseroad - thanks for your help!

**Pull request closed without merging**